File: 5-Problem-A-Day/174. Dungeon Game.java
class Solution {
    public int calculateMinimumHP(int[][] dungeon) {  
        int m = dungeon.length;
        int n = dungeon[0].length;

        // hp[y][x]: min hp required to reach bottom right (P).

        // int[][] hp = new int[m + 1][n + 1];
        int[] hp = new int[n + 1];
        Arrays.fill(hp, Integer.MAX_VALUE);
        hp[n - 1] = 1;
        // for (int[] row : hp) Arrays.fill(row, Integer.MAX_VALUE);
        // hp[m][n - 1] = hp[m - 1][n] = 1;

        for (int y = m - 1; y >= 0; --y)
            for (int x = n - 1; x >= 0; --x)
                hp[x] = Math.max(1, Math.min(hp[x], hp[x + 1]) - dungeon[y][x]);

        return hp[0];
    }
}

File: 2021/174. Dungeon Game.cpp
class Solution {
public:
    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        int m = dungeon.size(), n = dungeon[0].size();
        int dp[m][n];
        memset(dp, sizeof(dp), 0);        
        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {                
                dp[i][j] = (i == m - 1)
                    ? ((j == n - 1) ? dungeon[i][j] : dp[i][j + 1] + dungeon[i][j])
                    : ((j == n - 1) ? dp[i + 1][j] + dungeon[i][j] : max(dp[i + 1][j], dp[i][j + 1]) + dungeon[i][j]);
                if (dp[i][j] > 0) {
                    dp[i][j] = 0;
                }
            }
        }
        return 1 - dp[0][0];
    }
};